File: src/main/java/engine/graphics/query/GLTimerQuery.java
package engine.graphics.query;

import renderer.OpenGLContext;

import static org.lwjgl.opengl.GL15.GL_QUERY_RESULT;
import static org.lwjgl.opengl.GL15.glGenQueries;
import static org.lwjgl.opengl.GL33.*;

public class GLTimerQuery implements GLQuery<Float> {

    private static GLTimerQuery instance;
    public static GLTimerQuery getInstance() {
        if(instance == null) {
            instance = new GLTimerQuery();
        }
        return instance;
    }

    private final int start;
    private final int end;
    private volatile boolean finished = false;
    private boolean started;

    public GLTimerQuery() {
        start = glGenQuery();
        end = glGenQuery();
    }

    private int glGenQuery() {
        return OpenGLContext.getInstance().calculate( () -> glGenQueries());
    }

    @Override
    public void begin() {
        finished = false;
        OpenGLContext.getInstance().execute(() -> {
            glQueryCounter(start, GL_TIMESTAMP);
        }, true);
        started = true;
    }

    @Override
    public void end() {
        if(!started) {
            throw new IllegalStateException("Don't end a query before it was started!");
        }
        OpenGLContext.getInstance().execute(() -> {
            glQueryCounter(end, GL_TIMESTAMP);
        }, true);
        finished = true;
    }


    public long getTimeTaken() {
        return getEndTime() - getStartTime();
    }

    @Override
    public int getQueryToWaitFor() {
        return start;
    }

    public long getStartTime() {
        return OpenGLContext.getInstance().calculate( () -> glGetQueryObjectui64(start, GL_QUERY_RESULT));
    }

    public long getEndTime() {
        return OpenGLContext.getInstance().calculate( () -> glGetQueryObjectui64(end, GL_QUERY_RESULT));
    }

    @Override
    public Float getResult() {
        if(!finished) { throw new IllegalStateException("Don't query result before query is finished!"); }
        while(!resultsAvailable()) {
        }
        return getTimeTaken() / 1000000f;
    }
}

File: engine/src/main/kotlin/de/hanno/hpengine/graphics/query/GLTimerQuery.kt
package de.hanno.hpengine.graphics.query

import de.hanno.hpengine.graphics.GpuContext
import org.lwjgl.opengl.GL15
import org.lwjgl.opengl.GL33

class GLTimerQuery(gpuContext: GpuContext) : GLQuery<Float?> {
    private val gpuContext: GpuContext
    override val queryToWaitFor: Int
    private val end: Int

    @Volatile
    private var finished = false
    private var started = false
    private fun glGenQuery(): Int {
        return gpuContext.onGpu { GL15.glGenQueries() }
    }

    override fun begin(): GLTimerQuery? {
        finished = false
        gpuContext.onGpu {
            GL33.glQueryCounter(queryToWaitFor, GL33.GL_TIMESTAMP)
            Unit
        }
        started = true
        return this
    }

    override fun end() {
        check(started) { "Don't end a query before it was started!" }
        gpuContext.onGpu {
            GL33.glQueryCounter(end, GL33.GL_TIMESTAMP)
            Unit
        }
        finished = true
    }

    val timeTaken: Long
        get() = endTime - startTime
    val startTime: Long
        get() = gpuContext.onGpu { GL33.glGetQueryObjectui64(queryToWaitFor, GL15.GL_QUERY_RESULT) }
    val endTime: Long
        get() = gpuContext.onGpu { GL33.glGetQueryObjectui64(end, GL15.GL_QUERY_RESULT) }
    override val result: Float
        get() {
            check(finished) { "Don't query result before query is finished!" }
            while (!resultsAvailable(gpuContext)) {
            }
            return timeTaken / 1000000f
        }

    init {
        this.gpuContext = gpuContext
        queryToWaitFor = glGenQuery()
        end = glGenQuery()
    }
}